File: app/src/main/java/com/arturogutierrez/openticator/domain/issuer/IssuerDecoratorFactory.java
package com.arturogutierrez.openticator.domain.issuer;

import android.content.Context;
import com.arturogutierrez.openticator.R;
import com.arturogutierrez.openticator.di.PerActivity;
import com.arturogutierrez.openticator.domain.issuer.model.Issuer;
import java.util.ArrayList;
import java.util.EnumMap;
import java.util.List;
import javax.inject.Inject;

@PerActivity
public class IssuerDecoratorFactory {

  private final Context context;
  private EnumMap<Issuer, String> names;
  private EnumMap<Issuer, Integer> images;

  @Inject
  public IssuerDecoratorFactory(Context context) {
    this.context = context;
    this.names = new EnumMap<>(Issuer.class);
    this.images = new EnumMap<>(Issuer.class);

    initialize();
  }

  private void initialize() {
    names.put(Issuer.AWS, context.getString(R.string.aws));
    images.put(Issuer.AWS, R.drawable.icon_account_aws);
    names.put(Issuer.DIGITALOCEAN, context.getString(R.string.digital_ocean));
    images.put(Issuer.DIGITALOCEAN, R.drawable.icon_account_digitalocean);
    names.put(Issuer.BITCOIN, context.getString(R.string.bitcoin));
    images.put(Issuer.BITCOIN, R.drawable.icon_account_bitcoin);
    names.put(Issuer.DROPBOX, context.getString(R.string.dropbox));
    images.put(Issuer.DROPBOX, R.drawable.icon_account_dropbox);
    names.put(Issuer.EVERNOTE, context.getString(R.string.evernote));
    images.put(Issuer.EVERNOTE, R.drawable.icon_account_evernote);
    names.put(Issuer.FACEBOOK, context.getString(R.string.facebook));
    images.put(Issuer.FACEBOOK, R.drawable.icon_account_facebook);
    names.put(Issuer.GITHUB, context.getString(R.string.github));
    images.put(Issuer.GITHUB, R.drawable.icon_account_github);
    names.put(Issuer.GOOGLE, context.getString(R.string.google));
    images.put(Issuer.GOOGLE, R.drawable.icon_account_google);
    names.put(Issuer.MICROSOFT, context.getString(R.string.microsoft));
    images.put(Issuer.MICROSOFT, R.drawable.icon_account_microsoft);
    names.put(Issuer.SLACK, context.getString(R.string.slack));
    images.put(Issuer.SLACK, R.drawable.icon_account_slack);
    names.put(Issuer.WORDPRESS, context.getString(R.string.wordpress));
    images.put(Issuer.WORDPRESS, R.drawable.icon_account_wordpress);
  }

  public List<IssuerDecorator> create(List<Issuer> issuers) {
    List<IssuerDecorator> issuerDecorators = new ArrayList<>(issuers.size());
    for (Issuer issuer : issuers) {
      IssuerDecorator issuerDecorator = create(issuer);
      issuerDecorators.add(issuerDecorator);
    }
    return issuerDecorators;
  }

  private IssuerDecorator create(Issuer issuer) {
    String name = names.get(issuer);
    Integer image = images.get(issuer);
    return new IssuerDecorator(issuer, name, image);
  }
}

File: app/src/main/java/com/arturogutierrez/openticator/domain/issuer/IssuerDecoratorFactory.kt
package com.arturogutierrez.openticator.domain.issuer

import android.content.Context
import com.arturogutierrez.openticator.R
import com.arturogutierrez.openticator.di.PerActivity
import com.arturogutierrez.openticator.domain.issuer.model.Issuer
import java.util.*
import javax.inject.Inject

@PerActivity
class IssuerDecoratorFactory @Inject constructor(val context: Context) {

  private val names: Map<Issuer, String>
  private val images: Map<Issuer, Int>

  init {
    names = mapOf(
        Issuer.AWS to context.getString(R.string.aws),
        Issuer.DIGITALOCEAN to context.getString(R.string.digital_ocean),
        Issuer.BITCOIN to context.getString(R.string.bitcoin),
        Issuer.DROPBOX to context.getString(R.string.dropbox),
        Issuer.EVERNOTE to context.getString(R.string.evernote),
        Issuer.FACEBOOK to context.getString(R.string.facebook),
        Issuer.GITHUB to context.getString(R.string.github),
        Issuer.GOOGLE to context.getString(R.string.google),
        Issuer.MICROSOFT to context.getString(R.string.microsoft),
        Issuer.SLACK to context.getString(R.string.slack),
        Issuer.WORDPRESS to context.getString(R.string.wordpress)

    )
    images = mapOf(
        Issuer.AWS to R.drawable.icon_account_aws,
        Issuer.DIGITALOCEAN to R.drawable.icon_account_digitalocean,
        Issuer.BITCOIN to R.drawable.icon_account_bitcoin,
        Issuer.DROPBOX to R.drawable.icon_account_dropbox,
        Issuer.EVERNOTE to R.drawable.icon_account_evernote,
        Issuer.FACEBOOK to R.drawable.icon_account_facebook,
        Issuer.GITHUB to R.drawable.icon_account_github,
        Issuer.GOOGLE to R.drawable.icon_account_google,
        Issuer.MICROSOFT to R.drawable.icon_account_microsoft,
        Issuer.SLACK to R.drawable.icon_account_slack,
        Issuer.WORDPRESS to R.drawable.icon_account_wordpress)
  }

  fun create(issuers: List<Issuer>): List<IssuerDecorator> {
    val issuerDecorators = ArrayList<IssuerDecorator>(issuers.size)
    for (issuer in issuers) {
      val issuerDecorator = create(issuer)
      issuerDecorators.add(issuerDecorator)
    }
    return issuerDecorators
  }

  private fun create(issuer: Issuer): IssuerDecorator {
    val name = names[issuer]!!
    val image = images[issuer]!!
    return IssuerDecorator(issuer, name, image)
  }
}
